File: leetcode/1-100/27-remove-element/bench/two_pointer.py
"""Benchmark workload — two-pointer in-place O(n) Remove Element.

Algorithmic mirror of bench/two_pointer.kara. See ../README.md § Benchmarks
for the choice of N, K and the LCG ~50%-match input.
"""

from __future__ import annotations


def remove_element(nums: list[int], length: int, val: int) -> int:
    k = 0
    for i in range(length):
        if nums[i] != val:
            nums[k] = nums[i]
            k += 1
    return k


def main() -> None:
    n = 2_000_000
    val = 0

    original = [0] * n
    state = 1
    for i in range(n):
        state = (state * 1103515245 + 12345) % 2147483648
        if (state // 65536) % 2 == 1:
            original[i] = i + 1
        else:
            original[i] = 0

    workspace = [0] * n

    sum_result = 0
    for _ in range(10):
        for p in range(n):
            workspace[p] = original[p]
        k = remove_element(workspace, n, val)
        sum_result += k + workspace[k - 1]
    print(sum_result)


if __name__ == "__main__":
    main()

File: leetcode/1-100/27-remove-element/bench/two_pointer.rs
//! Benchmark workload — two-pointer in-place O(n) Remove Element.
//!
//! Algorithmic mirror of bench/two_pointer.kara and bench/two_pointer.py. See
//! ../README.md § Benchmarks for the choice of N, K and the LCG ~50%-match
//! input.

fn remove_element(nums: &mut [i64], len: i64, val: i64) -> i64 {
    let mut k = 0i64;
    for i in 0..len {
        if nums[i as usize] != val {
            nums[k as usize] = nums[i as usize];
            k += 1;
        }
    }
    k
}

fn main() {
    const N: usize = 2_000_000;
    let val: i64 = 0;

    let mut original: Vec<i64> = vec![0; N];
    let mut state: i64 = 1;
    for i in 0..N {
        state = (state * 1103515245 + 12345) % 2147483648;
        if (state / 65536) % 2 == 1 {
            original[i] = i as i64 + 1;
        } else {
            original[i] = 0;
        }
    }

    let mut workspace: Vec<i64> = vec![0; N];

    let mut sum: i64 = 0;
    for _ in 0..10 {
        for p in 0..N {
            workspace[p] = original[p];
        }
        let k = remove_element(&mut workspace, N as i64, val);
        sum += k + workspace[(k - 1) as usize];
    }
    println!("{}", sum);
}
